docs/references/security/rust-canister-development-security-best-practices.md

```diff
@@ -183,43 +183,157 @@ A canister could be rendered unusable so it could never be upgraded again e.g. d
 
--   Generally, there can be bugs when data is not persisted when the developer expected it to be.
+Note that in Rust, from Rust CDK version 0.5.1, any local variables still go out of scope on panic. This prevents some of the issues above. 
 
 #### Recommendation
 
--   [Don’t panic after `await`](https://mmapped.blog/posts/01-effective-rust-canisters.html#panic-await) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
+- Watch the [community conversation on Security Best Practices](https://www.youtube.com/watch?v=PneRzDmf_Xw&list=PLuhDt1vhGcrez-f3I0_hvbwGZHZzkZ7Ng&index=2&t=4s) which shows a concrete example of an issue as described here.
 
--   [Don’t lock shared resources across await boundaries](https://mmapped.blog/posts/01-effective-rust-canisters.html#dont-lock) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
+- [Don’t lock shared resources across await boundaries](https://mmapped.blog/posts/01-effective-rust-canisters.html#dont-lock) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
 
--   See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
+- [Don’t panic after `await`](https://mmapped.blog/posts/01-effective-rust-canisters.html#panic-await) (from [Effective Rust Canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html))
 
--   For context: [IC interface spec on message execution](../ic-interface-spec#message-execution)
+- See also: "Inter-canister calls" section in [How to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister)
 
-### Be aware that state may change during inter-canister calls
+- For context: [IC interface spec on message execution](../ic-interface-spec#message-execution)
+
+### Be aware that there is no reliable message ordering
 
 #### Security Concern
 
-Messages (but not entire calls) are processed atomically. This can lead to security issues, such as:
+As described in the Message Execution Basics above, messages (but not entire calls) are processed atomically. This can lead to a variety of severe security issues such as:
+
+* _Time-of-check time-of-use issues._ These occur when some condition on global state is checked before an inter-canister call is made, and then wrongly assuming the condition to still hold when the call returns. For example, one might check if there is sufficient balance on some account, then issue two inter-canister calls (e.g. making transfers). When the second inter-canister call starts, it is possible that the condition checked initially no longer holds, because other ledger transfers may have happened before the callback of the first call is executed (see also Property 3 above). 
```

Comment:
```suggestion
* _Time-of-check time-of-use issues._ These occur when some condition on global state is checked before an inter-canister call, and then wrongly assuming the condition still holds when the call returns. For example, one might check if there is sufficient balance on some account before making a transfer as part of the callback message. When the second inter-canister call starts, it is possible that the condition which was checked initially no longer holds, because other ledger transfers may have happened before the callback of the first call is executed (see also Property 3 above). 
```